### ISSUE TITLE
integrations: CML

### ISSUE THREAD

**Issue by daavoo** [iterative/dvclive#91]
From a high level user perspective, I think it would be interesting to have some kind of integration between `dvclive` and `CML`.

Just like there is already a custom `dvclive` behavior when used along with `DVC` (i.e. generating `checkpoints`) maybe a similar approach could be used when `dvclive` is also used along with `CML`.

I was thinking on a use case like `dvclive` taking care of using `cml-send-comment` inside `next_step` if `CML` is detected. 

Conceptually i think that it would be something similar to the existing `make_checkpoint`:

https://github.com/iterative/dvclive/blob/master/dvclive/metrics.py#L105  

This might be just an specific use case of https://github.com/iterative/dvclive/issues/90

**Comment by dberenbaum** [comment]
This seems like a general need for integration between cml and dvc checkpoints. Do you see any functionality as being specific to dvclive?

**Comment by daavoo** [comment]
> This seems like a general need for integration between cml and dvc checkpoints. Do you see any functionality as being specific to dvclive?

Given the discussion so far in https://github.com/iterative/dvc/issues/6182, it looks like the `exp push` integration won't require any changes from the `dvclive` side. 

I probably didn't express it right but I was also taking about an additional integration, which will be something like using `cml-send-comment` in the context of a long running experiment (i.e. deep learning) as a way to periodically report the status of the training. 

From an user perspective what I would want is, by using `dvclive` in an environment where CML is installed, getting a P.R. comment reporting the `dvclive` metrics *summary* **each** time an epoch is finished (which would translate to each time `dvclive.next_step` is called). 

This is a feature (sending the metrics to somewhere) that I have been repeatedly implementing through workarounds when using other *ML Loggers* and I though that ot could be nice to have it working out of the box with `dvclive<>CML`.

The feature probably doesn't make too much sense in the context of training a lot of short experiments and it's probably not a deal breaker for attracting new users.

**Comment by dberenbaum** [comment]
> The feature probably doesn't make too much sense in the context of training a lot of short experiments and it's probably not a deal breaker for attracting new users.

Yes, this could be really useful, and absolutely dvclive should address issues for long-running experiments! https://github.com/iterative/dvc/issues/6182 will at least make it possible to pull the latest checkpoint and see the associated metrics and plots, but we could make the experience smoother and better integrate cml.

Seems like this has 2 requirements:

1. Generate some default report to send.
2. Trigger `cml-send-comment` at checkpoints or some other point in code (should it be coupled with checkpoints/next_step?).

**Comment by shcheklein** [comment]
@iterative/cml I know that you have been discussing (and even prototyping) something like this? Could you share your thoughts, outcomes?

**Comment by pared** [comment]
@dberenbaum I think we could even start with pretty printing metrics to some env var that CML could pick up and show in the build log.

**Comment by casperdcl** [comment]
I'd say generic callback support (#90) would be best. The user can then run `cml-send-comment --update` per-checkpoint.

Vaguely related (though this is automatic rather than manual): auto-push per checkpoint https://github.com/iterative/dvc/issues/6182)

**Comment by DavidGOrtega** [comment]
> From a high level user perspective, I think it would be interesting to have some kind of integration between dvclive and CML.

That would be awesome. I think that the best integration would be if dvclive would work as tensorboard. Something that you can launch and give you a url in return. We wrapped that under cml-tensorboard-dev.

> @iterative/cml I know that you have been discussing (and even prototyping) something like this? Could you share your thoughts, outcomes?

I implemented something like that in ```cml-publish```. The idea was to initiate a daemon when you used cml-publish to watch for file changes of the file that you published with ```cml-publish```. If the daemon see a new change republishes the file again in the same url. It works like badges, so the images in the report are always refreshed. The daemon was a very simple piece of software using [chokidar](https://github.com/paulmillr/chokidar).

**Comment by dberenbaum** [comment]
That sounds cool, @DavidGOrtega! Definitely cleaner than a separate report for each checkpoint. In this case, can dvclive republish to the same url at each step without needing a daemon?

**Comment by DavidGOrtega** [comment]
Yes, it could. Something that actually @pared also liked when I shared the idea in general and gave the same feedback. 
Its not implemented in CML, its a local experiment like the extension that we did

**Comment by daavoo** [comment]
I'm currently working on what I consider the minimum viable integration between DVCLive and CML

- DVCLive P.R. https://github.com/iterative/dvclive/pull/215
- Integration example https://github.com/iterative/test-dvclive-cml
- Current result https://github.com/iterative/test-dvclive-cml/commit/8460322df5a586d89882367310c1a4d534202c24

The current prototype doesn't depend on DVC (or DVC checkpoints) but still works with both. 

It is focused on using `cml send-comment` and `cml publish` to share the progress of a training stage **while the python process it's being run**.

- If CML is installed, on each step update (i.e. `next_step()`), DVCLive generates a markdown report using the metrics already logged. 
- The report includes a table (uses `tabulate`) and rendered images for the plots (uses `cml publish`).
- Calls `cml send-comment` once the report is ready.

It currently uses `subprocess.run` to call `cml` commands. 

It only works for posting multiple comments (1 per step) because of  https://github.com/iterative/cml/issues/888
I still have to try to set up a workflow supporting "P.R. comments" instead of "commit comments".

@iterative/cml I would like to hear some thoughts about the downsides of raw `subprocess.run` v.s. [potential python bindings](https://github.com/iterative/cml/issues/828) to see how "blocker" those bindings are.

**Comment by DavidGOrtega** [comment]
@daavoo I can't follow up the links.

I did a wrapper for python called CMLpy. Basically it exposes all the CML commands.
I just only need to do the installer. We should sync. I'm going to send you an invite

**Comment by daavoo** [comment]
> @daavoo I can't follow up the links.

Updated the permissions

**Comment by dberenbaum** [comment]
@daavoo Could you give me permission, too 🙏 ?

**Comment by daavoo** [comment]
> @daavoo Could you give me permission, too 🙏 ?

Done

**Comment by dberenbaum** [comment]
Looks great! Should `report` be set by config file/environment variable? It seems better to not have to touch the code to toggle this option (similar to `html`).

**Comment by DavidGOrtega** [comment]
@dberenbaum @daavoo I have invited you to pycml we should invite everyone

**Comment by dberenbaum** [comment]
> I'd say generic callback support (#90) would be best. The user can use that to run `cml-send-comment --update` per-checkpoint.
> 
> Vaguely related (though this is automatic rather than manual): auto-push per checkpoint [iterative/dvc#6182](https://github.com/iterative/dvc/issues/6182)

What about this? Can dvclive have some built-in CML callback but also allow users to add their own?

**Comment by dberenbaum** [comment]
@daavoo:

I'd say the important requirements from the dvclive user perspective here are:
1. This should be as close as possible to the same report that gets generated locally in html when running a script with dvclive. Since one is in html and the other in markdown, they can't be identical, but it should seem like the same report whether generated locally or via cml.
2. I like the metrics table, but this means we should try to find a way to include it in the local html also.
3. Minimal extra dependencies, and whatever is needed is installed on the default cml image.

@DavidGOrtega:

Base on your earlier comments about how to integrate, I'm wondering if it would be possible to reverse this integration? Could cml host this report, watch for changes to it, and update the comment accordingly?

**Comment by DavidGOrtega** [comment]
> Base on your earlier comments about how to integrate, I'm wondering if it would be possible to reverse this integration? Could cml host this report, watch for changes to it, and update the comment accordingly?

There are some quirks @dberenbaum the biggest issue is the GH image cache. While redoing the log in every epoch using `send-comment --update` is right now solved and similar to other logging platform

**Comment by daavoo** [comment]
> What about this? Can dvclive have some built-in CML callback but also allow users to add their own?

It makes sense to me to offer flexibility to the users but only with the limited scope of sharing the report at the end of each step. 

Given that DVCLive integrations are already a callback in the ML Frameworks, it feels over-complicated to support custom callbacks inside `Live`.

**Comment by dberenbaum** [comment]
> It makes sense to me to offer flexibility to the users but only with the limited scope of sharing the report at the end of each step.
> 
> Given that DVCLive integrations are already a callback in the ML Frameworks, it feels over-complicated to support custom callbacks inside `Live`.

Good point. The problem here and in other issues like #90 and #204 is that users can already add similar callbacks directly in their model training code, so integrations/callbacks only add value if we are using the info specifically captured by DVCLive.

**Comment by dberenbaum** [comment]
I mentioned this to @0x2b3bfa0, but I wanted to post it in the issue to get more feedback. Happy to open an issue in CML, but wanted to discuss here first.

> So to follow up on your thoughts about CML-DVCLive integration, it seems possible to have a more modular approach to this instead of needing a Python API if:
> 1. There is some option like `cml send-comment --publish` that publishes all referenced images from the markdown and converts the markdown to reference the published images.
> 2. There is some file watcher daemon like `cml send-comment --watch` that performs the action whenever the file is updated.
> 
> With those options, something like `cml send-comment --publish --watch dvclivereport.md` could intermittently send comments whenever DVCLive updates the report.

**Comment by 0x2b3bfa0** [comment]
> _I think you mean a new feature request like cml publish --watch <plot.png> (i.e. static URL), constantly updated image._

I think he doesn't quite mean that. GitHub caches images; updating them implies replacing the link by a different one.

**Comment by 0x2b3bfa0** [comment]
> _would technically need a subprocess file-watcher that also checked if the file was mid-way through being updated. Implementation is a bit advanced, but it's definitely great for end-user UX (single CLI flag, no API code)._

Sounds interesting, but would require users to run something in the background 🤔

**Comment by 0x2b3bfa0** [comment]
![1598352521_128552_1598354440_sumario_normal](https://user-images.githubusercontent.com/11387611/167158062-0ab98513-cdca-46dd-a1a8-2331eeb1f749.jpg)

**Comment by casperdcl** [comment]
Ah. That's yet another use case. Trying to keep track of ideas:

1. report per-epoch metrics
2. publish per-epoch plots
3. auto-parse, publish & substitute local markdown links (is that a common use case?)

or is there something else?

**Comment by dberenbaum** [comment]
I see it all as one use case: publish per-epoch reports (see #90). The typical DVCLive report includes metrics and plots, and it would be great to publish the entire report as a CML comment.

To achieve CML integration, DVCLive can either call CML directly, or it can generate a markdown report that can be useful anywhere (locally, in CML, and possibly for publishing elsewhere). If DVCLive goes with the option to just generate markdown, we need some way to convert from local markdown file to something that can be published in CML.

I also see it as potentially useful for CML generally. Instead of:

```
          cat metrics.txt >> report.md
          cml-publish plot.png --md >> report.md
          cml-send-comment report.md
```

You can do:

```
          cat metrics.txt >> report.md
          cat plot.png >> report.md
          cml-send-comment --publish report.md
```

This makes workflows less dependent on CML-specific commands, and it enables users to do things like have Python scripts that generate complex markdown and then publish them as comments with one command.

**Comment by dberenbaum** [comment]
See also https://github.com/iterative/example-repos-dev/blob/a58b1220ef76c5b498929b2c017aa7a9e61eaa3c/example-get-started/code/.github/workflows/cml.yaml:

```
          echo "# CML Report" > report.md
          echo "## Plots" >> report.md
          dvc plots diff $PREVIOUS_REF workspace \
            --show-vega --targets evaluation/plots/precision_recall.json > vega.json
          vl2svg vega.json prc.svg
          cml publish prc.svg --title "Precision & Recall" --md >> report.md
          dvc plots diff $PREVIOUS_REF workspace \
            --show-vega --targets evaluation/plots/confusion_matrix.json > vega.json
          vl2svg vega.json confusion.svg
          cml publish confusion.svg --title "Confusion Matrix" --md >> report.md
          echo "## Metrics and Params" >> report.md
          echo "### $PREVIOUS_REF → workspace" >> report.md
          dvc exp diff $PREVIOUS_REF --show-md >> report.md
          cml send-comment report.md
```

It would be great if DVC and/or DVCLive can provide utilities to more simply generate the content, and have CML take care of making all the content available in a comment. It might look something like:

```
          echo "# CML Report" > report.md
          echo "## Plots" >> report.md
          # Add support to output plots directly to png
          dvc plots diff $PREVIOUS_REF workspace \
            --png --targets evaluation/plots/precision_recall.json \
            --title "Precision & Recall" >> report.md
          dvc plots diff $PREVIOUS_REF workspace \
            --png --targets evaluation/plots/confusion_matrix.json \
            --title "Confusion Matrix" >> report.md
          echo "## Metrics and Params" >> report.md
          echo "### $PREVIOUS_REF → workspace" >> report.md
          dvc exp diff $PREVIOUS_REF --show-md >> report.md
          cml send-comment --publish report.md
```

Even better would be if something like `dvc exp diff --plots --show-md` generated the entire `report.md`.

**Comment by 0x2b3bfa0** [comment]
See also https://github.com/iterative/cml/pull/1026

**Comment by casperdcl** [comment]
Just to confirm:

- #90 (user callback for `cml send-comment --update`)
- #215 (DVCLive calls `cml send-comment --update`)

are not needed. Instead, the proposal https://github.com/iterative/cml/issues/1036 is

- DVCLive option to write markdown (instead of html)

```bash
## current
while true; do
  if [[ -f report.md ]]; then
    for img in $(extract_images --file=report.md); do
      sed -i s/$img/$(cml publish $img)/ report.md
    done
    cml send-comment --update report.md
  fi
  sleep 5
done &
python train.py --per-epoch-report=report.md

## proposed
cml report watch --publish report.md &
python train.py --per-step-dvclive-output=report.md

## dvc way, no code modification, only dvc exp params
cml report watch --publish report.md &
dvc exp run -S dvclive_output=report.md
```

@daavoo can you add this option to DVCLive? Note that `report.md` can contain local images `![](./image.png)` - CML will automagically `publish` them :)

**Comment by daavoo** [comment]
> @daavoo can you add this option (markdown instead of html) to DVCLive?

Will do. Have to work on some pre-requisites https://github.com/iterative/dvc-render/issues/62
